ethcore/src/state/mod.rs

```diff
@@ -952,8 +952,78 @@ impl<B: Backend> State<B> {
 		}))
 	}
 
+	#[cfg(feature="to-pod-full")]
+	/// Populate a PodAccount map from this state.
+	/// Warning this is not for real time use.
+	/// Use of this method requires FatDB mode to be able 
+	/// to iterate on accounts.
+	pub fn to_pod_full(&self) -> PodState {
+
+		assert!(self.checkpoints.borrow().is_empty());
+		assert!(self.factories.trie.is_fat());
+
+		let mut result = BTreeMap::new();
+
+		let trie = match self.factories.trie.readonly(self.db.as_hashdb(), &self.root) {
+			Ok(trie) => trie,
+			_ => {
+				trace!(target: "state", "to_pod_full: Couldn't open the DB");
+				return PodState::from(result);
+			}
+		};
+
+		let iter = match trie.iter() {
+			Ok(iter) => iter,
+			_ => return PodState::from(result),
+		};
+
+		// put trie in cache 
+		for item in iter {
+			if let Ok((addr, _dbval)) = item {
+				let address = Address::from_slice(&addr);
+				let _ = self.require(&address, true);
+			}
+		}
+
+		// Resolve missing part
+		for (add, opt) in self.cache.borrow_mut().iter_mut() {
+			if let Some(ref mut acc) = opt.account {
+				let pod_account = self.account_to_pod_account(acc, add);
+				result.insert(add.clone(), pod_account);
+			}
+		}
+
+		PodState::from(result)
+	}
+
+	fn account_to_pod_account(&self, account: &mut Account, address: &Address) -> PodAccount {
+		let mut pod_storage = BTreeMap::new();
+		let addr_hash = account.address_hash(address);
+		let accountdb = self.factories.accountdb.readonly(self.db.as_hashdb(), addr_hash);
+
+		if account.code_size().is_none() {
+			let _ = account.cache_code(accountdb.as_hashdb());
+		}
+
+		account.storage_root()
+			.map(|root| self.factories.trie.readonly(accountdb.as_hashdb(), &root)
+				.map(|trie|	trie.iter()
+					.map(|iter| iter.for_each(|o_kv|{
+						if let Ok((key, val)) = o_kv {
+							pod_storage.insert(H256::from(&key[..]), H256::from(&val[..]));
+						}
+					}
+					))));
+
+		let mut pod_account = PodAccount::from_account(&account);
+		// cached one first
+		pod_storage.append(&mut pod_account.storage);
```

Comment:
Please use `self.storage_at` instead. The issue is that what we have in trie and what we have in cache might differ, and we need to use the value in cache.

Reply:
> Reuse of `storage_at` was not easy in this case (could not iterate and
use the method at the same time (refcell mutable borrow panics) so
keeping code as is.

Would you mind to try this:

* Collect all storage items from trie first in a `Vec<H256>`.
* Call `storage_at` at the end of the round.

Not sure whether that would solve the lifetime issue.

Reply:
Yes this will works (but I kinda wanted to avoid that :) ), I will switch to this version for the sake of using similar code.

Reply:
> Please use self.storage_at instead. The issue is that what we have in trie and what we have in cache might differ, and we need to use the value in cache.

Doesn't the merge ('append') of both storage do that? (pod_account got values from storage_change replacing value read from trie). Can value from 'original_storage_cache' of 'Account' differs from trie value (this field is lost in the process) ? 

Switching to 'storage_at' anyway.

Reply:
@cheme Ah sorry. Yeah you're right. `PodAccount::from` would only print things on `storage_changes`, so using `State::storage_at` won't be sufficient for our case. I think the current method works.